Implement opaque closures

Keno opened the pull request:
This is the end result of the design process in #31253.

 # Overview

This PR implements a new kind of closure, called an `opaque closure`.
It is designed to be complimenatry to the existing closure mechanism
and makes some different trade offs. The motivation for this mechanism
comes primarily from closure-based AD tools, but I'm expecting it will
find other use cases as well. From the end user perspective, opaque
closures basically behave like regular closures, expect that they
are introduced by adding the `@opaque` macro (not part of this PR,
but will be added after). In front of the existing closure. In
particular, all scoping, capture, etc. rules are identical. For
such user written closures, the primary difference is in the
performance characteristics. In particular:

1) Passing an opaque closure to a high order function will specialize
   on the argument and return types of the closure, but not on the
   closure identity. (This also means that the opaque closure will not
   be eligible for inlining into the higher order function, unless the
   inliner can see both the definition and the call site).
2) The optimizer is allowed to modify the capture environment of the
   opaque closure (e.g. dropping unused captures, or reducing `Box`ed
   values back to value captures).

The `opaque` part of the naming comes from the notion that semantically,
nothing is supposed to inspect either the code or the capture environment
of the opaque closure, since the optimizer is allowed to choose any value
for these that preserves the behavior of calling the opaque closure itself.

 # Motivation
 ## Optimization across closure boundaries

Consider the following situation (type annotations are inference
results, not type asserts)
```
function foo()
    a = expensive_but_effect_free()::Any
    b = something()::Float64
    ()->isa(b, Float64) ? return nothing : return a
end
```

now, the traditional closure mechanism will lower this to:

```
struct ###{T, S}
   a::T
   b::S
end
(x::###{T,S}) = isa(x.b, Float64) ? return nothing : return x.a
function foo()
    a = expensive_but_effect_free()::Any
    b = something()::Float64
    new(a, b)
end
```

the problem with this is apparent: Even though (after inference),
we know that `a` is unused in the closure (and thus would be
able to delete the expensive call were it not for the capture),
we may not delete it, simply because we need to satisfy the full
capture list of the closure. Ideally, we would like to have a mechanism
where the optimizer may modify the capture list of a closure in
response to information it discovers.

 ## Closures from Casette transforms

Compiler passes like Zygote would like to generate new closures
from untyped IR (i.e. after the frontend runs) (and in the future
potentially typed IR also). We currently do not have a great mechanism
to support this. This provides a very straightforward implementation
of this feature, as opaque closures may be inserted at any point during
the compilation process (unlike types, which may only be inserted
by the frontend).

 # Mechanism

The primary concept introduced by this PR is the `OpaqueClosure{A<:Tuple, R}`
type, constructed, by the new `Core._opaque_closure` builtin, with
the following signature:

```
    _opaque_closure(argt::Type{<:Tuple}, lb::Type, ub::Type, source::CodeInfo, captures...)

Create a new OpaqueClosure taking arguments specified by the types `argt`. When called,
this opaque closure will execute the source specified in `source`. The `lb` and `ub`
arguments constrain the return type of the opaque closure. In particular, any return
value of type `Core.OpaqueClosure{argt, R} where lb<:R<:ub` is semantically valid. If
the optimizer runs, it may replace `R` by the narrowest possible type inference
was able to determine. To guarantee a particular value of `R`, set lb===ub.
```

Captures are available to the CodeInfo as `getfield` from Slot 1
(referenced by position).

 # Examples

I think the easiest way to understand opaque closures is look through
a few examples. These make use of the `@opaque` macro which isn't
implemented yet, but makes understanding the semantics easier.
Some of these examples, in currently available syntax can be seen
in test/opaque_closure.jl

```
oc_trivial() = @opaque ()::Any->1
@show oc_trivial() # ()::Any->◌
@show oc_trivial()() # 1

oc_inf() = @opaque ()->1
 # Int return type is inferred
@show oc_inf() # ()::Int->◌
@show oc_inf()() # 1

function local_call(b::Int)
    f = @opaque (a::Int)->a + b
    f(2)
end

oc_capture_opt(A) = @opaque (B::typeof(A))->ndims(A)*B
@show oc_capture_opt([1; 2]) # (::Vector{Int},)::Vector{Int}->◌
@show sizeof(oc_capture_opt([1; 2]).env) # 0
@show oc_capture_opt([1 2])([3 4]) # [6 8]
```

Keno committed:
Implement opaque closures

This is the end result of the design process in #31253.

 # Overview

This PR implements a new kind of closure, called an `opaque closure`.
It is designed to be complimenatry to the existing closure mechanism
and makes some different trade offs. The motivation for this mechanism
comes primarily from closure-based AD tools, but I'm expecting it will
find other use cases as well. From the end user perspective, opaque
closures basically behave like regular closures, expect that they
are introduced by adding the `@opaque` macro (not part of this PR,
but will be added after). In front of the existing closure. In
particular, all scoping, capture, etc. rules are identical. For
such user written closures, the primary difference is in the
performance characteristics. In particular:

1) Passing an opaque closure to a high order function will specialize
   on the argument and return types of the closure, but not on the
   closure identity. (This also means that the opaque closure will not
   be eligible for inlining into the higher order function, unless the
   inliner can see both the definition and the call site).
2) The optimizer is allowed to modify the capture environment of the
   opaque closure (e.g. dropping unused captures, or reducing `Box`ed
   values back to value captures).

The `opaque` part of the naming comes from the notion that semantically,
nothing is supposed to inspect either the code or the capture environment
of the opaque closure, since the optimizer is allowed to choose any value
for these that preserves the behavior of calling the opaque closure itself.

 # Motivation
 ## Optimization across closure boundaries

Consider the following situation (type annotations are inference
results, not type asserts)
```
function foo()
    a = expensive_but_effect_free()::Any
    b = something()::Float64
    ()->isa(b, Float64) ? return nothing : return a
end
```

now, the traditional closure mechanism will lower this to:

```
struct ###{T, S}
   a::T
   b::S
end
(x::###{T,S}) = isa(x.b, Float64) ? return nothing : return x.a
function foo()
    a = expensive_but_effect_free()::Any
    b = something()::Float64
    new(a, b)
end
```

the problem with this is apparent: Even though (after inference),
we know that `a` is unused in the closure (and thus would be
able to delete the expensive call were it not for the capture),
we may not delete it, simply because we need to satisfy the full
capture list of the closure. Ideally, we would like to have a mechanism
where the optimizer may modify the capture list of a closure in
response to information it discovers.

 ## Closures from Casette transforms

Compiler passes like Zygote would like to generate new closures
from untyped IR (i.e. after the frontend runs) (and in the future
potentially typed IR also). We currently do not have a great mechanism
to support this. This provides a very straightforward implementation
of this feature, as opaque closures may be inserted at any point during
the compilation process (unlike types, which may only be inserted
by the frontend).

 # Mechanism

The primary concept introduced by this PR is the `OpaqueClosure{A<:Tuple, R}`
type, constructed, by the new `Core._opaque_closure` builtin, with
the following signature:

```
    _opaque_closure(argt::Type{<:Tuple}, lb::Type, ub::Type, source::CodeInfo, captures...)

Create a new OpaqueClosure taking arguments specified by the types `argt`. When called,
this opaque closure will execute the source specified in `source`. The `lb` and `ub`
arguments constrain the return type of the opaque closure. In particular, any return
value of type `Core.OpaqueClosure{argt, R} where lb<:R<:ub` is semantically valid. If
the optimizer runs, it may replace `R` by the narrowest possible type inference
was able to determine. To guarantee a particular value of `R`, set lb===ub.
```

Captures are available to the CodeInfo as `getfield` from Slot 1
(referenced by position).

 # Examples

I think the easiest way to understand opaque closures is look through
a few examples. These make use of the `@opaque` macro which isn't
implemented yet, but makes understanding the semantics easier.
Some of these examples, in currently available syntax can be seen
in test/opaque_closure.jl

```
oc_trivial() = @opaque ()::Any->1
@show oc_trivial() # ()::Any->◌
@show oc_trivial()() # 1

oc_inf() = @opaque ()->1
 # Int return type is inferred
@show oc_inf() # ()::Int->◌
@show oc_inf()() # 1

function local_call(b::Int)
    f = @opaque (a::Int)->a + b
    f(2)
end

oc_capture_opt(A) = @opaque (B::typeof(A))->ndims(A)*B
@show oc_capture_opt([1; 2]) # (::Vector{Int},)::Vector{Int}->◌
@show sizeof(oc_capture_opt([1; 2]).env) # 0
@show oc_capture_opt([1 2])([3 4]) # [6 8]
```

vtjnash commented:
As a mildly big change, I’m hoping to get a chunk of time next week to review, amidst any v1.6 branching questions and bugs. Can you make sure `-compile=all` is working in the mean time? And briefly highlight what parts of the bigger plan are completed here and which things are future work (so I don’t focus too much on the wrong things). Seems like possibly also a small test set, but maybe they really are that simple and lack many edge cases or errors?

Keno commented:
The intent of this PR is to get the semantics right and add the minimum possible stub code in all the right places to make the various cases I want work. I'm sure that there are lots of missing edge cases (for example I know I haven't done anything yet with opaque closures that take a vararg), but I think it's in a good enough state to get onto master and continue iterating there. It's not really in the state where I would recommend people immediately jump on it. I think the primary thing to focus on in the review would be the semantics of the new builtin, the lattice element, etc. Everything else is just a matter of fixing some corner cases, but if we get the semantics wrong, we're obviously in trouble.

oxinabox commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1326,14 +1354,14 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                 if tchanged(rt, frame.bestguess)
                     # new (wider) return type for frame
-                    frame.bestguess = tmerge(frame.bestguess, rt)
+                    frame.bestguess = frame.bestguess === NOT_FOUND ? rt : tmerge(frame.bestguess, rt)
```
Would it be reasonable to add this into `tmerge` such that: `tmerge(::NotFound, x) = x`

Keno replied:
tmerge is very hot and this isn't technically part of the lattice, but rather a sentinel value only used in a few places. It used to be in tmerge, but @vtjnash moved it out a few months ago.

vtjnash replied:
How did bestguess manage to get initialized to the wrong value? (I'm not sure why we even have it anymore, instead of using Union{}, but seems like ⊑ still has the code to handle it, for some reason that I don't remember from #36596 dcc0696971a100dbc63f90b8789632631dd1bfef, also)

oxinabox commented on base/compiler/inferencestate.jl:
```diff
@@ -112,7 +121,7 @@ mutable struct InferenceState
             Vector{Tuple{InferenceState,LineNum}}(), # cycle_backedges
             Vector{InferenceState}(), # callers_in_cycle
-            #=parent=#nothing,
+            #=parent=#nothing, #= has_opaque_closures =# false,
```
Why does the style change from annoating on right (old line 113 and 114) to annotating inline?
```suggestion
            nothing, # parent
            false, # has_opaque_closures
```

oxinabox commented on base/compiler/inferencestate.jl:
```diff
@@ -242,6 +251,7 @@ end
 
 # temporarily accumulate our edges to later add as backedges in the callee
 function add_backedge!(li::MethodInstance, caller::InferenceState)
+    caller.linfo !== nothing || return # don't add backedges to opauqe closures
```
```suggestion
    caller.linfo !== nothing || return # don't add backedges to opaque closures
```

vtjnash replied:
Should this be a TODO XXX? I'd assume that OpaqueClosures do have edges (ie. can call functions), so we need to be able to propagate those back to their caller

oxinabox commented on base/compiler/inferencestate.jl:
```diff
@@ -242,6 +251,7 @@ end
 
 # temporarily accumulate our edges to later add as backedges in the callee
 function add_backedge!(li::MethodInstance, caller::InferenceState)
+    caller.linfo !== nothing || return # don't add backedges to opauqe closures
     isa(caller.linfo.def, Method) || return # don't add backedges to toplevel exprs
```
Is it work adding documentairy helper functions
```
is_toplevel(::InferenceState)
is_opaque_closure(::InferenceState)
```

oxinabox commented on base/compiler/optimize.jl:
```diff
@@ -34,7 +34,7 @@ struct InliningState{S <: Union{EdgeTracker, Nothing}, T <: Union{InferenceCache
 
 mutable struct OptimizationState
-    linfo::MethodInstance
+    linfo::Union{MethodInstance, Nothing}
```
Would it add clarity to define a new singleton to use a signal something is a OpaqueClosure, 
rather than using `Nothing` for this?
Or is using `nothing` a sitatution that in theory could occur more widely than just for OpaqueClosures?

Or from another direction, should there be a `OpaqueClosureOptimizationState` that doens't have this field?

Keno replied:
Yeah, that might be cleaner, but honestly this whole thing needs a refactor, but that's for another PR.

oxinabox commented on base/compiler/optimize.jl:
```diff
@@ -311,6 +311,13 @@ function statement_cost(ex::Expr, line::Int, src::CodeInfo, sptypes::Vector{Any}
                 ftyp = argextype(farg, src, sptypes, slottypes)
             end
         end
+        # Give calls to OpaqueClosures zero cost. The plan is for these to be a single
```
It seems like everything below here should be a seperate function.
(Or rather a collection of seperate methods).

Is the reason it isn't to do with compilation costs?

Keno replied:
No, this could be refactored.

oxinabox commented on base/compiler/ssair/inlining.jl:
```diff
@@ -533,6 +550,12 @@ function batch_inline!(todo::Vector{Pair{Int, Any}}, ir::IRCode, linetable::Vect
         for ((old_idx, idx), stmt) in compact
             if old_idx == inline_idx
                 argexprs = copy(stmt.args)
+                if isa(item, InliningTodo) && item.mi === nothing
```
this again makes me this we should use a different marker than `nothing` to indicate that this is a opaque closure.

Keno replied:
There are MethodInstances for opaque closures at this stage, inlining just doesn't support that yet.

oxinabox commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1165,6 +1223,10 @@ function assemble_inline_todo!(ir::IRCode, state::InliningState)
                 ir.stmts[idx][:inst] = quoted(calltype.val)
                 continue
             end
+            # Refuse to inline OpaqueClosures we can't see otherwise, to preserve the
```
should this be either:
```suggestion
            # Otherwise: refuse to inline OpaqueClosures we can't see, to preserve the
```

or:
```suggestion
            # Refuse to inline OpaqueClosures we can't see, to preserve the
```

Or am i not understanding what is going on here?

Keno replied:
This comment seems to have gotten lost.

oxinabox commented on base/compiler/tfuncs.jl:
```diff
@@ -1368,6 +1368,24 @@ function array_builtin_common_nothrow(argtypes::Array{Any,1}, first_idx_idx::Int
+    argt, argt_exact = instanceof_tfunc(arg)
+    lbt, lb_exact = instanceof_tfunc(lb)
+    if !lb_exact
+    lbt = Union{}
```
```suggestion
        lbt = Union{}
```

oxinabox commented on base/compiler/typeinfer.jl:
```diff
@@ -173,6 +176,49 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult, val
+    argtypes = Any[argt.parameters...]
+    pushfirst!(argtypes, clos.env)
+    if isdispatchtuple(argt)
+        # If we don't need to track specializations, just infer this here
+        # right now.
```
```suggestion
        # If we don't need to track specializations, just infer this here right now.
```

Julia's line limit is 92 character so this fits on one line

JeffBezanson commented on src/jltypes.c:
```diff
@@ -2468,8 +2469,17 @@ void jl_init_types(void) JL_GC_DISABLED
+        jl_perm_symsvec(4, "env", "ci", "fptr1", "fptr"),
+        jl_svec(4, jl_any_type, jl_code_info_type, pointer_void, pointer_void), 0, 0, 4)->name->wrapper;
+    jl_opaque_closure_typename = ((jl_datatype_t*)jl_unwrap_unionall((jl_value_t*)jl_opaque_closure_type))->name;
+    jl_compute_field_offsets((jl_datatype_t*)jl_unwrap_unionall((jl_value_t*)jl_opaque_closure_type));
```
I think this could be defined in boot.jl.

Keno replied:
For the moment yes, but I'm anticipating giving these special layout to make the fields truly invisible and let LLVM resize them for the capture list, which wouldn't be available from boot.jl.

JeffBezanson commented on src/julia.h:
```diff
@@ -351,6 +351,19 @@ struct _jl_method_instance_t {
     uint8_t inInference; // flags to tell if inference is running on this object
 };
 
+// YACK - Yet another kind of closure.
```
```suggestion
// OpaqueClosure
```

JeffBezanson commented on src/julia.h:
```diff
@@ -351,6 +351,19 @@ struct _jl_method_instance_t {
+    union {
+        jl_value_t *code;
+        jl_code_info_t *source;
+        jl_method_t *method;
```
When are each of these cases used?

JeffBezanson committed:
add front-end support for OpaqueClosure

oxinabox commented:
A thing @vtjnash  and I discussed in the multithreading call was that `Threads.@spawn` could potentially be changed to interally create Opaque Closures.
Which could have some benifits, and would not violate API since noone can really look at that closure anyway.

vtjnash commented:
> I think it's in a good enough state to get onto master and continue iterating there

Ah, okay. I thought you were trying to merge this into v1.6. Makes more sense if we're actually giving this some time on master to finish and mature. I'm hoping we can branch that off ASAP, so that we can start merging more feature work and getting experience with great new advancements like this.

Keno committed:
Address some review

Keno committed:
Some better optimizations for OpaqueClosure

Keno committed:
Implement code_typed reflection for OpaqueClosures

Keno committed:
Backport some fixes

Keno committed:
Some extra backports

oscardssmith commented:
@Keno I've been trying to use this, and I'm getting some errors. Can you confirm that this transformation should work?

```
function stext(idx::Int)
	f = io -> print(io, "hi")
    return f
end

function stext_opaque(idx::Int)
	source = Meta.lower(@__MODULE__, :(print(io, "hi"))).args[1]
	f = Core._opaque_closure(Tuple{<:IO}, Nothing, Nothing, source, idx)
    return f
end
```
When I try to run `stext_opaque`, I get a whole screen of errors.

Keno commented:
There is frontend support now, so you'll probably want to try that:
```
julia> using Base: @opaque

julia> function stext(idx::Int)
           return @opaque io->print(io, "hi")
       end
stext (generic function with 1 method)

julia> stext(1)
(::Any)::Nothing->◌

julia> ans(stdout)
hi
```
That said, performance here may not yet reflect the performance characteristics of the final feature.

oscardssmith commented:
Thanks for making the frontend work! This is now working. It currently is a little slower than a regular closure largely due to extra allocations.

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -34,15 +34,33 @@ function normalize(@nospecialize(stmt), meta::Vector{Any})
 end
 
-function convert_to_ircode(ci::CodeInfo, code::Vector{Any}, coverage::Bool, nargs::Int, sv::OptimizationState)
+function add_opaque_closure_argtypes!(argtypes, t)
```
Seems like a few methods still need type annotations in the Core.Compiler module. I don't want to mark each case.

Aside: also seems like this particular one might be unused? but github might just be hiding the right file diff

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -34,15 +34,33 @@ function normalize(@nospecialize(stmt), meta::Vector{Any})
+end
+
+
+function convert_to_ircode(ci::CodeInfo, code::Vector{Any}, coverage::Bool, nargs::Int, sv::OptimizationState, slottypes=sv.slottypes, stmtinfo=sv.stmt_info)
```
nit: line length (github is acting unwilling to show me the whole line)

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -65,6 +67,9 @@ end
 function ssa_inlining_pass!(ir::IRCode, linetable::Vector{LineInfoNode}, state::InliningState, propagate_inbounds::Bool)
     # Go through the function, performing simple ininlingin (e.g. replacing call by constants
```
```suggestion
    # Go through the function, performing simple ininling (e.g. replacing call by constants
```

jstrube replied:
inlining?

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -295,6 +300,14 @@ function finish_cfg_inline!(state::CFGInliningState)
+    if isexpr(stmt′, :call) && is_known_call(stmt′, getfield, ir, Any[]) && stmt′.args[2] === Argument(1)
+        stmt′.args[1] = Core.getfield_opaque_env
+    end
+end
```
```suggestion
    nothing
end
```

vtjnash commented on base/opaque_closure.jl:
```diff
@@ -0,0 +1,29 @@
+@noinline function (y::Core.OpaqueClosure{A, R})(args...) where {A,R}
+    typeassert(args, A)
+    ccall(y.fptr1, Any, (Any, Ptr{Any}, Int), y, Any[args...], length(args))::R
+end
```
This code probably should probably now be moved to C (where `jl_apply_generic` is) as the invoke handler for objects of this type, where we can better deal with race-conditions and updating the pointer after reoptimization and calling convention changes and such (and hopefully optimizations too). I think that'll actually even make the `typeassert` disappear into the dynamic dispatch caching operation on the previous line.

vtjnash commented on base/opaque_closure.jl:
```diff
@@ -0,0 +1,29 @@
+end
+
+
+# @opaque macro goes here
```
```suggestion
"""
@opaque macro docstring goes here
"""
```

vtjnash commented on base/opaque_closure.jl:
```diff
@@ -0,0 +1,29 @@
+Core._opaque_closure
+
+function show(io::IO, oc::Core.OpaqueClosure{A, R}) where {A, R}
+    types = map(@nospecialize(T)->Expr(:(::), T), A.parameters)
```
`A` might not have `.parameters`, in general?

vtjnash commented on base/opaque_closure.jl:
```diff
@@ -0,0 +1,29 @@
+
+# @opaque macro goes here
+macro opaque(ex)
+    Expr(:opaque_closure, esc(ex))
```
```suggestion
    return esc(Expr(:opaque_closure, ex))
```

nit: this is actually slightly simpler IR (though a non-functional change), since we'll immediately elide the macro-scope/escape pair during macroexpand instead of waiting for lowering

vtjnash commented on base/reflection.jl:
```diff
@@ -1165,6 +1168,24 @@ function code_typed_by_type(@nospecialize(tt::Type);
+    if isa(clos.ci, CodeInfo)
+        code = copy(clos.ci)
+        debuginfo === :none && remove_linenums!(code)
+        return [code=>code.rettype]
```
```suggestion
        return Any[Pair{CodeInfo,Any}(code, code.rettype)]
```

vtjnash commented on src/codegen.cpp:
```diff
@@ -5738,15 +5824,27 @@ static std::pair<std::unique_ptr<Module>, jl_llvm_functions_t>
+    bool is_opaque_closure = false;
+    if (jl_is_method(lam->def.method)) {
+        ctx.nargs = lam->def.method->nargs;
+    } else {
```
style nit:
```suggestion
    }
    else {
```

vtjnash commented on src/interpreter.c:
```diff
@@ -44,9 +44,9 @@ extern void JL_GC_ENABLEFRAME(interpreter_state*) JL_NOTSAFEPOINT;
 // behavior and terminates the exploration
 #define JL_GC_PUSHFRAME(frame,n)     \
-  JL_CPPALLOCA(frame, sizeof(*frame)+((n) * sizeof(jl_value_t*)));                  \
+  JL_CPPALLOCA(frame, sizeof(*frame)+((n+3) * sizeof(jl_value_t*)));                  \
```
This should be a separate PR, and can land immediately. Since it's just for clang-sa, probably doesn't need to be backported?
```suggestion
  JL_CPPALLOCA(frame, sizeof(*frame)+(((n)+1) * sizeof(jl_value_t*)));               \
  ((void**)&frame[1])[0] = NULL;                                                    \
  memset(&((void**)&frame[1])[1], 0, (n)*sizeof(jl_value_t*));                      \
  _JL_GC_PUSHARGS(&((void**)&frame[1])[1], n);
```

Keno committed:
Refactor JL_GC_PUSHFRAME to internalize layout assumptions

Right now callers of JL_GC_PUSHFRAME need to be aware that the
roots start three pointers after the end of the struct. That
seems like an implementation detail of the macro, so refactor the
macro to also take the assignment location for the locals instead.
In the process fix an error where the gc-analyzer version of this
code assumed a different layout.

Keno committed:
Fix review nits

Keno committed:
Switch opaque closure creation to IR form only

As requested in review.

Keno committed:
Fix opaque closure show method

Keno committed:
Fix tag number

Keno commented:
I've addressed the review comments and switched the builtin to an IR form instead. I'll plan to merge in a couple days and keep iterating on master, but if there's additional comments, do keep them coming.

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -1033,6 +1082,36 @@ function process_simple!(ir::IRCode, todo::Vector{Pair{Int, Any}}, idx::Int, sta
+    # Inlining OpaqueClosure that don't have any specializations
+    if sig.ft ⊑ Core.OpaqueClosure
+        callee = stmt.args[1]
+        if isa(callee, SSAValue) && isexpr(ir.stmts[callee.id][:inst], :new_opaque_closure) &&
+                length(ir.stmts[callee.id][:inst].args) >= 5 &&
+                isa(ir.stmts[callee.id][:inst].args[4], OpaqueClosureIdx)
+            ir′ = ir.opaques[(ir.stmts[callee.id][:inst].args[4]::OpaqueClosureIdx).n]::IRCode
+            push!(todo, idx=>InliningTodo(nothing, ResolvedInliningSpec(ir′, linear_inline_eligible(ir′)), true))
+            return nothing
```
DRY style: values should not be recomputed unnecessarily (and long lines often split by introducing named intermediates)
```suggestion
        if isa(callee, SSAValue)
             callee_inst = ir.stmts[callee.id][:inst]
             if isexpr(callee_inst, :new_opaque_closure) && length(callee_inst.args) >= 5
                idx = callee_inst.args[4]
                if isa(idx, OpaqueClosureIdx)
                    ir′ = ir.opaques[idx.n]::IRCode
                    todo = InliningTodo(nothing, ResolvedInliningSpec(ir′, linear_inline_eligible(ir′)), true))
                    push!(todo, idx => todo)
                    return nothing
                end
            end
```

vtjnash replied:
Why `>= 5`? That should have a comment. (We know that `length(callee_inst.args) >= 4` is guaranteed by the IR validator, so we don't need to assert that anywhere else)

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -994,6 +1024,22 @@ function inline_invoke!(ir::IRCode, idx::Int, sig::Signature, invoke_data::Invok
+            stmt.args[2] = stmt.args[3] = widenconst(tmerge(lb, m.unspecialized.cache.rettype))
+        end
+    end
+end
```
```suggestion
    nothing
end
```

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -994,6 +1024,22 @@ function inline_invoke!(ir::IRCode, idx::Int, sig::Signature, invoke_data::Invok
+        lb, exact = instanceof_tfunc(lbt)
+        exact || return
+        # Narrow opaque closure type
+        if isa(calltype.ci, OptimizationState)
```
```suggestion
        source = calltype.ci
        if isa(source, OptimizationState)
        ...
```
(same nit: avoiding recomputing values)

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -994,6 +1024,22 @@ function inline_invoke!(ir::IRCode, idx::Int, sig::Signature, invoke_data::Invok
+        exact || return
+        # Narrow opaque closure type
+        if isa(calltype.ci, OptimizationState)
+            stmt.args[2] = stmt.args[3] = widenconst(tmerge(lb, calltype.ci.src.rettype))
```
Why do we not use the inferred type `widenconst(calltype)`?

vtjnash commented on base/compiler/ssair/inlining.jl:
```diff
@@ -295,6 +300,15 @@ function finish_cfg_inline!(state::CFGInliningState)
+    # If this is an access into the opaque closure environment,
+    # rewrite it to a call to `getfield_opaque_env`
+    if isexpr(stmt′, :call) && is_known_call(stmt′, getfield, ir, Any[]) && stmt′.args[2] === Argument(1)
+        stmt′.args[1] = Core.getfield_opaque_env
```
```suggestion
        stmt′.args[1] = GlobalRef(Core, :getfield_opaque_env)
```

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -118,24 +152,37 @@ function slot2reg(ir::IRCode, ci::CodeInfo, nargs::Int, sv::OptimizationState)
-    preserve_coverage = coverage_enabled(sv.mod)
-    ir = convert_to_ircode(ci, copy_exprargs(ci.code), preserve_coverage, nargs, sv)
+function compact_all!(ir::IRCode)
+    length(ir.stmts) == 0 && return ir
```
style nit:
```suggestion
    isempty(ir.stmts) && return ir
```
Is this actually legal IR? I though it had to contain `:return`, at least

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -34,15 +34,35 @@ function normalize(@nospecialize(stmt), meta::Vector{Any})
+
+function convert_to_ircode(ci::CodeInfo, code::Vector{Any},
+                           coverage::Bool, nargs::Int, sv::OptimizationState,
+                           slottypes=sv.slottypes, stmtinfo=sv.stmt_info)
```
types on these?

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -58,7 +78,21 @@ function convert_to_ircode(ci::CodeInfo, code::Vector{Any}, coverage::Bool, narg
+            if isa(ssat.ci, OptimizationState)
+                opaque_ir = make_ir(ssat.ci.src, 0, ssat.ci)
+                push!(opaques, opaque_ir)
+                stmt.head = :new_opaque_closure
```
```suggestion
```

vtjnash commented on base/compiler/ssair/driver.jl:
```diff
@@ -34,15 +34,35 @@ function normalize(@nospecialize(stmt), meta::Vector{Any})
 end
 
-function convert_to_ircode(ci::CodeInfo, code::Vector{Any}, coverage::Bool, nargs::Int, sv::OptimizationState)
+function add_opaque_closure_argtypes!(argtypes::Vector{Any}, @nospecialize(t))
+    dt = unwrap_unionall(t)
+    dt1 = unwrap_unionall(dt.parameters[1])
+    if isa(dt1, TypeVar) || isa(dt1.parameters[1], TypeVar)
+        push!(argtypes, Any)
+    else
+        TT = dt1.parameters[1]
+        if isa(TT, Union)
+            TT = tuplemerge(TT.a, TT.b)
+        end
+        for p in TT.parameters
+            push!(argtypes, rewrap_unionall(p, t))
+        end
+    end
+end
```
```suggestion
    nothing
end
```

vtjnash replied:
```suggestion
```
(appears to be dead code)

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1015,6 +1017,34 @@ function abstract_call_known(interp::AbstractInterpreter, @nospecialize(f),
     return abstract_call_gf_by_type(interp, f, argtypes, atype, sv, max_methods)
 end
 
+function abstract_call_opaque_closure(interp::AbstractInterpreter, clos::PartialOpaque, argtypes::Vector{Any}, sv::InferenceState)
```
To review placeholder: start here (jwn)

vtjnash replied:
```suggestion
function abstract_call_opaque_closure(interp::AbstractInterpreter, closure::PartialOpaque, argtypes::Vector{Any}, sv::InferenceState)
```

vtjnash commented on base/compiler/typeinfer.jl:
```diff
@@ -301,6 +301,9 @@ function CodeInstance(result::InferenceResult, @nospecialize(inferred_result::An
         elseif isa(result.result, PartialStruct)
             rettype_const = (result.result::PartialStruct).fields
             const_flags = 0x2
+        elseif isa(result.result, PartialOpaque)
```
To review placeholder: start here (jwn)

vtjnash reviewed (commented):
About 95% reviewed, but github is struggling now, so will finish up a bit later. You might want to open a new PR when you update to speed up render.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -203,6 +203,8 @@ function const_prop_profitable(@nospecialize(arg))
             isconstType(b) && return true
             const_prop_profitable(b) && return true
         end
+    elseif isa(arg, PartialOpaque)
```
I thought the major benefit of `OpaqueClosure` is that you don't have to aggressively specialize on the value. This seems to be doing precisely the opposite, and aggressively forcing repeated inference on any user of them.

vtjnash commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1015,6 +1017,34 @@ function abstract_call_known(interp::AbstractInterpreter, @nospecialize(f),
+        pushfirst!(nargtypes, clos.env)
+        result = InferenceResult(Core.OpaqueClosure, nargtypes)
+        state = InferenceState(result, copy(clos.ci), false, interp)
+        typeinf_local(interp, state)
```
This code effectively claims that an OpaqueClosure cannot call itself. However, that would seem to imply it cannot be passed as an argument, as otherwise we can build a y-combinator with it and invalidate this claim. It also implies that it can't call any recursive function that could call it (also a faulty assumption, for the same reason). I think that implies we need the majority of the same logic as abstract_call_method here. Probably by refactoring some helper functions?

Since we don't optimize much yet anyways, I'd suggested that currently it should treat all opaque objects as potential derivatives of the same template.

vtjnash commented on base/compiler/typeinfer.jl:
```diff
@@ -381,6 +384,64 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
+
+    argt = unwrap_unionall(clos.t).parameters[1]
+
+    argtypes = Any[argt.parameters...]
```
Not all subtypes of Tuple have `.parameters` (maybe use `fieldtype`?)

vtjnash commented on base/compiler/typeinfer.jl:
```diff
@@ -381,6 +384,64 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
+    end
+end
+
+function finish_rettype!(@nospecialize(rt), mod::Module, interp::AbstractInterpreter)
+    if isa(rt, PartialOpaque)
+        finish_opaque_closure!(rt, mod, interp)
+    elseif isa(rt, PartialStruct)
+        for field in rt.fields
+            if isa(field, PartialStruct) || isa(field, PartialOpaque)
+                finish_rettype!(field, mod, interp)
+            end
+        end
+    end
+end
```
```suggestion
```
seems dead?

vtjnash commented on base/compiler/typeinfer.jl:
```diff
@@ -381,6 +384,64 @@ function cache_result!(interp::AbstractInterpreter, result::InferenceResult)
 end
 
+function finish_opaque_closure!(clos::PartialOpaque, mod::Module, interp::AbstractInterpreter)
+    # Infer with the most general types possible, so that optimization has
+    # access to the result.
```
The implication in `finish` is that inference is already done, since otherwise we might still have an unresolved cycle that must be converged first before we can annotate the correct (sufficiently wide) types. Thus, this is not a point in the graph where it is valid to look to look at IPO information.

Sixzero commented:
I just built this branch and ran some tests to see what opaque closures can do:
https://gist.github.com/Sixzero/149e391b41879ac10d0952d6e0e97fab

I see some allocations hurting the performance for small functions, but it is all right, what bothers me more is that LLVM can't optimalize the opaque functions further and discarding the allocations for the arrays in return statements.

As @Keno said in his video, we could go hard on opaque function optimizations, and trade little more compilation time for little more optimized code in the case of opaque closures. 

I believe there needs to be a way to get better performance, probably this is all about generating MLIR instead of LLVM-IR? Sorry if this question is a weak connected to the issue.

Keno commented on src/codegen.cpp:
```diff
@@ -4480,6 +4490,98 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
+            return jl_cgval_t();
+        }
+        if (jl_is_code_info(source.constant)) {
+            // TODO: Emit this inline and outline it late using LLVM's coroutine
+            // support.
```
coroutine support would allow llvm passes to optimize the capture list. We could write our own version of it, but I think the existing one would work fine.

Keno commented on src/opaque_closure.c:
```diff
@@ -0,0 +1,64 @@
+}
+
+
+JL_DLLEXPORT jl_method_t* jl_mk_opaque_closure_method(jl_module_t *def_mod)
```
Yes, it's not quite a real `Method`, since it doesn't get inserted in the global method list either and also shouldn't really appear in any `.sig` key-ed caches, since the `.sig` field has a different meaning for these methods. I think this needs to be part of a larger refactor of what it means to be a method, since I would also like to have alternative method tables for offload targets. However, for the purposes of this PR, I would like to keep this as is and keep iterating. I can disallow these methods in dump.c for now until we figure out a better way to deal with them.

Keno commented on src/codegen.cpp:
```diff
@@ -4480,6 +4490,98 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
+            return jl_cgval_t();
+        }
+        jl_cgval_t *argv = (jl_cgval_t*)alloca(sizeof(jl_cgval_t) * nargs);
+        for (size_t i = 0; i < nargs; ++i) {
+            argv[i] = emit_expr(ctx, args[i]);
+        }
+        const jl_cgval_t &argt = argv[0];
+        const jl_cgval_t &ub = argv[2];
+        const jl_cgval_t &source = argv[3];
```
> `argv[0]` is required to be an actual literal?

I don't want to make that restriction in general.

Keno commented on src/codegen.cpp:
```diff
@@ -4480,6 +4490,98 @@ static jl_cgval_t emit_expr(jl_codectx_t &ctx, jl_value_t *expr, ssize_t ssaval)
+            }
+            li->specTypes = (jl_value_t*)jl_apply_tuple_type_v(jl_svec_data(sig_args), nsig);
+            li->def.module = ctx.module;
+            std::tie(closure_m, closure_decls) = emit_function(li, closure_src,
+                ub.constant, ctx.emission_context);
+            jl_merge_module(ctx.f->getParent(), std::move(closure_m));
```
I'm not convinced that's a good idea. We should never recurse into type inference here and it does not have an independent identity by which we could refer to it.

Keno commented:
May have taken 6 months, but almost every thing that was in here is on master now.

alok commented:
@Keno in a related vein, any eta for a public/beta version of diffractor.jl? i'd love to try it out

Keno commented:
A few weeks, once it's actually in a tryable form.
